File: impls/scala/step1_read_print.scala
import reader.tokenize

object step1_read_print {
  // read
  def READ(str: String): Any = {
    reader.read_str(str)
  }

  // eval
  def EVAL(ast: Any, env: String): Any = {
    ast
  }

  // print
  def PRINT(exp: Any): String = {
    printer._pr_str(exp, true)
  }

  // repl
  def main(args: Array[String]) = {
    val REP = (str: String) => {
      PRINT(EVAL(READ(str), ""))
    }

    var line:String = null
    while ({line = readLine("user> "); line != null}) {
      try {
        println(REP(line))
      } catch {
        case e : Exception => {
          println("Error: " + e.getMessage)
          println("    " + e.getStackTrace.mkString("\n    "))
        }
      }
    }
  }
}

// vim: ts=2:sw=2

File: impls/rust/step1_read_print.rs
#![allow(non_snake_case)]

extern crate fnv;
extern crate itertools;
extern crate regex;

mod readline;
#[macro_use]
#[allow(dead_code)]
mod types;
use crate::types::{MalRet, MalVal};
#[allow(dead_code)]
mod env;
mod printer;
mod reader;

// read
fn read(str: &str) -> MalRet {
    reader::read_str(str)
}

// eval
fn eval(ast: MalVal) -> MalRet {
    Ok(ast)
}

// print
fn print(ast: MalVal) -> String {
    ast.pr_str(true)
}

fn rep(str: &str) -> Result<String, MalVal> {
    let ast = read(str)?;
    let exp = eval(ast)?;
    Ok(print(exp))
}

fn main() {
    // `()` can be used when no completer is required

    // main repl loop
    while let Some(ref line) = readline::readline("user> ") {
        if !line.is_empty() {
            match rep(line) {
                Ok(ref out) => println!("{}", out),
                Err(ref e) => println!("Error: {}", e.pr_str(true)),
            }
        }
    }
    println!();
}
